### PR TITLE
[rmp-2] SortServiceManager

### PR DESCRIPTION
**Pull request by StanislavShchyrba**
- Added manager to SortService

- Synchronized binding SortService

- Refactoring code(All exceptions handling carried over to manager)

### CODE REVIEW

File: src/androidTest/java/com/example/sortingapp/ExampleInstrumentedTest.java [hunk 1]
```diff
@@ -1,4 +1,8 @@
+<<<<<<< HEAD:src/androidTest/java/com/example/sortingapp/ExampleInstrumentedTest.java
+package com.example.sortingapp;
+=======
 package com.example.usage;
+>>>>>>> f4bf8bf4ddac0cce6586beaf5deb0b0e087e29a6:src/androidTest/java/com/example/usage/ExampleInstrumentedTest.java
```
Comment (gutachkee) [inline src/androidTest/java/com/example/sortingapp/ExampleInstrumentedTest.java:5]: ?

File: src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.example.usage">
+    package="com.example.sortingapp">
```
Comment (gutachkee) [inline src/main/AndroidManifest.xml:3]: :+1:

File: src/main/cpp/generate.cpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#include <cstdlib>
+
+#include "generate.h"
```
Comment (gutachkee) [inline src/main/cpp/generate.cpp:3]: better names: "num_generator.h", "generator.h"

File: src/main/cpp/quicksort.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#ifndef QUICKSORT_H
+#define QUICKSORT_H
+
+namespace native_sorter {
+    void quickSort(int arr[], int low, int high);//wrapper
+
+    int partition(int arr[], int low, int high);
```
Comment (gutachkee) [inline src/main/cpp/quicksort.h:7]: this method is unused outside. Shouldn't be defined here.

File: src/main/java/com/example/sortingapp/MainActivity.java [hunk 5]
```diff
@@ -14,61 +11,49 @@
...
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
 
-        //Service creation and binding
-        Intent intent = new Intent(this, SortService.class);
-        bindService(intent, mConnection, Context.BIND_AUTO_CREATE);
+        new Thread(() -> {
+            isStartedSuccesfully = mSortManager.bind(MainActivity.this);
```
Comment (gutachkee) [inline src/main/java/com/example/sortingapp/MainActivity.java:44]: isStartedSuccesfully is unused. Use it in onClick() method: if isStartedSuccesfully == false -> show toast error.

ps. consider this name: isServiceBoundSuccesfully or isServiceBound

File: src/main/java/com/example/sortingapp/MainActivity.java [hunk 6]
```diff
@@ -14,61 +11,49 @@
...
+        }).start();
+
+        Toast.makeText(getApplicationContext(),
+                R.string.service_binded, Toast.LENGTH_SHORT).show();
```
Comment (gutachkee) [inline src/main/java/com/example/sortingapp/MainActivity.java:48]: Service is not bound at this point. Binding is done async. Remove this or place it right after `isStartedSuccesfully = mSortManager.bind(MainActivity.this);` line.

ps. if you move this to other thread, do not forget to post the toast showing to UI thread:
`new Handler(Looper.getMainLooper()).post(() -> showToastMessage(isServiceBound));`

File: src/main/java/com/example/sortingapp/MainActivity.java [hunk 7]
```diff
@@ -84,46 +69,50 @@ public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l)
...
             return;
         }
 
-        if (mNumbersAdapter.getDataSet().length == 0) {
-            Toast toast = Toast.makeText(getApplicationContext(),
-                    "No numbers to sort, press generate firstly", Toast.LENGTH_SHORT);
-            toast.show();
+        int[] sortedNumbersArray = null;
+
+        sortedNumbersArray = mSortManager.sort(mNumbersAdapter.getDataSet(), mSortingMethod);
```
Comment (gutachkee) [inline src/main/java/com/example/sortingapp/MainActivity.java:84]: make single line

File: src/main/java/com/example/sortmanager/SortServiceManager.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+
+public class SortServiceManager {
+    private static final long AWAIT_TIMEOUT_MILLS = 1000;
+
+    private CountDownLatch mLatch = new CountDownLatch(1);
+    private ISortService mSortService = null;
```
Comment (gutachkee) [inline src/main/java/com/example/sortmanager/SortServiceManager.java:27]:
```suggestion
    private volatile ISortService mSortService = null;
```

Create Trello ticket with "Learn about volatile fields"

File: src/test/java/com/example/sortingapp/ExampleUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+<<<<<<< HEAD:src/test/java/com/example/sortingapp/ExampleUnitTest.java
+package com.example.sortingapp;
+=======
+package com.example.usage;
+>>>>>>> f4bf8bf4ddac0cce6586beaf5deb0b0e087e29a6:src/test/java/com/example/usage/ExampleUnitTest.java
```
Comment (gutachkee) [inline src/test/java/com/example/sortingapp/ExampleUnitTest.java:5]: Remove unresolved conflicts.

File: src/test/java/com/example/usage/ExampleUnitTest.java [hunk 10]
```diff
@@ -1,4 +1,8 @@
+<<<<<<< HEAD:src/test/java/com/example/sortingapp/ExampleUnitTest.java
+package com.example.sortingapp;
+=======
 package com.example.usage;
+>>>>>>> f4bf8bf4ddac0cce6586beaf5deb0b0e087e29a6:src/test/java/com/example/usage/ExampleUnitTest.java
 
 import org.junit.Test;
```
Comment (gutachkee) [inline src/test/java/com/example/usage/ExampleUnitTest.java:8]: the same. remove.

**Review by gutachkee — CHANGES_REQUESTED** [review]
Better code quality. Good job. Keep going the good work.

few commens from me.

File: src/main/cpp/quicksort.h (hunk 4 above)
Comment (StanislavShchyrba) [reply src/main/cpp/quicksort.h:7]: partition is used in quickSort method. Where should I locate this method, inner anonymous namespace?

**Review by StanislavShchyrba — COMMENTED** [review]
Reverted unnecessary changes in unit tests.

**Review by gutachkee — APPROVED** [review]

**Pull request merged**